### PR TITLE
pkg/log: extend log configuration

### LINKED ISSUE #659: Optimize PLGD application logs of Kubernetes deployment.
## Feature Request
Optimize Logs of Kubernetes deployment. 

### What is the purpose of the feature?
Optimize logging by making better utilization of resources and address the following issues:
- Coap-Gateway and other PLGD app logs in debug mode are being written in stderr stream only.
- Coap-Gateway logs has entries (for notifications) that use more than 10 lines.

### What specific challenge does this feature solve?
By optimizing the logs to produce a single line per timestamp entry (for example putting it in a json) it allows log aggregator tools like Loki to better provide better log analysis (by reducing the amount of lines being produced in logs). It also helps if not all log entries are being sent to the same std stream. Errors should go to stderr, other log entries should go to stdout.

### PR DESCRIPTION
**Pull request by jkralik**
Examples:
COAP:
```jsonc
{
    "L": "DEBUG",
    "T": "2022-02-22T15:26:42.524269462Z",
    "M": "finished unary call from the device with code Changed",
    "startTime": "2022-02-22T15:26:42.481380231Z",
    "durationMs": 40.756,
    "request": {
        "body": {
            "accesstoken": "0ca0c61204a543abd3ee0ea2d16e829c",
            "authprovider": "plgd",
            "di": "b5a2a42e-b285-42f1-a36b-034c8fc8efd5"
        },
        "href": "/oic/sec/account",
        "method": "POST"
    },
    "response": {
        "body": {
            "accesstoken": "eyJhbGciOiJFUzI1NiIsImtpZCI6ImIwZWE1MGMxLTY3MGMtNWIwOC1iNjc3LWQzZTdlZmFlNWUwZiIsInR5cCI6IkpXVCJ9.eyJhdWQiOlsiaHR0cHM6Ly9sb2NhbGhvc3Q6MjAwMDkvIl0KLCJjbGllbnRfaWQiOiJ0ZXN0IiwiaHR0cHM6Ly9wbGdkLmRldi9kZXZpY2VJZCI6ImI1YTJhNDJlLWIyODUtNDJmMS1hMzZiLTAzNGM4ZmM4ZWZkNSIsImlhdCI6MTY0NTU0MzYwMgosImlzcyI6Imh0dHBzOi8vbG9jYWxob3N0OjIwMDA5LyIsInNjb3BlIjoib3BlbmlkIHByb2ZpbGUgZW1haWwgb2ZmbGluZV9hY2Nlc3MgcjoqIHc6KiIsInN1YiI6IjEifQ.y-eMN0RL7w4xxeVLHrc_8dugXfYOVMrKghky6EcueLTpgLMib8Src-e_Q_3d-QjSO6qRe2KrPmSdzjoRjy5ZYg",
            "expiresin": -1,
            "redirecturi": "",
            "refreshtoken": "refresh-token",
            "uid": "1"
        },
        "code": "Changed"
    },
    "protocol": "COAP"
}
```
GRPC:
```jsonc
{
    "L": "WARN",
    "T": "2022-02-21T20:29:11.522144612Z",
    "M": "finished unary call with code PermissionDenied",
    "response": {
        "code": "PermissionDenied",
        "error": "rpc error: code = PermissionDenied desc = cannot validate user access: rpc error: code = PermissionDenied desc = access denied"
    },
    "durationMs": 1.846,
    "protocol": "GRPC",
    "startTime": "2022-02-21T20:29:11.519007649Z",
    "request": {
        "correlationId": "8d08dd47-9b6e-4eb4-bbb9-3f1bcfa3f025",
        "deviceId": "a",
        "href": "/b",
        "jwt": {
            "sub": "1"
        },
        "method": "CreateResource",
        "service": "resourceaggregate.pb.ResourceAggregate"
    }
}
```
HTTP:
```jsonc
{
    "L": "DEBUG",
    "T": "2022-02-21T21:19:26.986032016Z",
    "M": "finished unary call with status code 200",
    "durationMs": 75.45,
    "request": {
        "href": "/api/v1/devices/6f1e8e14-e4fe-4282-672b-d6f89bd4a9f6/resource-links/switches",
        "jwt": {
            "sub": "1"
        },
        "method": "POST"
    },
    "response": {
        "code": 200
    },
    "startTime": "2022-02-21T21:19:26.909713334Z",
    "protocol": "HTTP"
}
```

Loki parser: https://grafana.com/docs/loki/latest/logql/log_queries/#json

https://github.com/plgd-dev/www/pull/48

### CODE REVIEW

File: coap-gateway/service/message/response.go [hunk 1]
```diff
@@ -25,7 +23,7 @@ func GetResponse(ctx context.Context, messagePool *pool.Pool, code codes.Code, t
 	}
 }
 
-func isTempError(err error) bool {
+func IsTempError(err error) bool {
 	switch {
 	case strings.Contains(err.Error(), "connect: connection refused"),
```
Comment (Danielius1922) [inline coap-gateway/service/message/response.go:29]: Maybe a TODO for the future: We could optimize this by using error.Is to avoid string comparison.

File: coap-gateway/service/signIn.go [hunk 2]
```diff
@@ -199,8 +200,8 @@ func subscribeAndValidateDeviceAccess(ctx context.Context, client *Client, owner
 	return client.server.ownerCache.OwnsDevice(ctx, deviceID)
 }
 
-func logSignInError(err error) {
-	log.Errorf("sign in error: %w", err)
+func logSignInError(err error) error {
```
Comment (Danielius1922) [inline coap-gateway/service/signIn.go]: `logSignInError` -> `signInError`, it no longer logs.

Comment (jkralik) [reply coap-gateway/service/signIn.go]: fixed

File: pkg/net/grpc/server/makeDefaultOptions.go [hunk 3]
```diff
@@ -2,27 +2,169 @@ package server
...
+// DefaultCodeToLevel is the default implementation of gRPC return codes and interceptor log level for server side.
+func DefaultCodeToLevel(code codes.Code) zapcore.Level {
+	lvl, ok := defaultCodeToLevel[code]
+	if ok {
+		return lvl
+	}
+	return zap.ErrorLevel
+}
+
+func setLogBasicLables(m map[string]interface{}, req interface{}) {
```
Comment (Danielius1922) [inline pkg/net/grpc/server/makeDefaultOptions.go]: typo Lables -> Labels

**Review by Danielius1922 — APPROVED** [review]

File: charts/plgd-hub/README.md [hunk 4]
```diff
@@ -34,8 +32,6 @@ global:
        # -- clientSecret or clientSecretFile
        clientSecret:
        #clientSecretFile:
-       # -- OAuth Scopes
```
Comment (ondrejtomcik) [inline charts/plgd-hub/README.md:37]: Why was this removed?

File: coap-gateway/service/log.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	"github.com/plgd-dev/go-coap/v2/mux"
+	"github.com/plgd-dev/go-coap/v2/tcp/message/pool"
+	coapgwMessage "github.com/plgd-dev/hub/v2/coap-gateway/service/message"
+	"github.com/plgd-dev/hub/v2/pkg/log"
+)
+
+const logRequestKey = "coap.request"
+const logRequestDeadlineKey = "coap.request.deadline"
+const logResponseKey = "coap.response"
+const logRequestToClientKey = "request to client"
```
Comment (ondrejtomcik) [inline coap-gateway/service/log.go]: What exactly is the request to the client? Who is the client? CoAP Gateway? In the context of our users who understands plgd hub as client interaction as:
`Client <-> gRPC GW <-> ... <-> RA <-> CoAP Gateway`
it's not understandable.

File: coap-gateway/service/log.go [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+func (client *Client) logClientRequest(req *mux.Message, resp *pool.Message) {
+	var rq *pool.Message
+	if req != nil {
+		tmp, err := client.server.messagePool.ConvertFrom(req.Message)
+		if err == nil {
+			rq = tmp
+		}
+	}
+	client.logRequest("request from client", rq, resp)
```
Comment (ondrejtomcik) [inline coap-gateway/service/log.go]: Same here.

**Comment by ondrejtomcik** [comment]
Could you please explain value `"coap.service":"coap-gateway"`in `"request from client"` log message? I didn't get it. Same for `"http.service": "http-gateway"`. Name it just `service`.

Update @jkralik : Just came to my mind that service is not needed at all. This is appended by the log aggregator, it is a duplicit information. It can me completely removed.

**Comment by ondrejtomcik** [comment]
Why you are using protocol specific keys like `"coap.start_time"`? It should be `request.start_time`, `request`, `response`, and so on.

TODO: camel case + part of the request/response log object.

**Comment by ondrejtomcik** [comment]
Why `plgd.deviceID`? What other deviceID will be there? use only `deviceID` key please.

**Comment by ondrejtomcik** [comment]
Both CoAP as well as HTTP have href. Use just `href` key. If I would want to filter HTTP gateway, I would combine `href` and `service` keys.

**Comment by ondrejtomcik** [comment]
Can you unify this? 
`"finished unary call with code Internal"`
`"finished unary call with status code 403"`

File: certificate-authority/config.yaml [hunk 7]
```diff
@@ -1,5 +1,11 @@
 log:
-  debug: false
+  level: info
+  encoding: json
+  stacktrace:
+    enabled: false
+    level: warn
```
Comment (ondrejtomcik) [inline certificate-authority/config.yaml:6]: It's too much. Keep only the `stacktrace` there. Why should I see the `stacktrace` in the warn, if it's only warning? I would change it to `displayStacktrace`

File: certificate-authority/config.yaml [hunk 8]
```diff
@@ -1,5 +1,11 @@
...
+  encoderConfig:
```
Comment (ondrejtomcik) [inline certificate-authority/config.yaml:7]: Again, it's too much in my opinion. I wouldn't allow configuration of this. Log aggregators can do whatever consumer wants.

File: coap-gateway/config.yaml [hunk 9]
```diff
@@ -1,6 +1,12 @@
 log:
-  debug: false
-  dumpCoapMessages: true
+  dumpBody: false
```
Comment (ondrejtomcik) [inline coap-gateway/config.yaml:2]: Request/response headers and query parameters are always dumped?

**Review by ondrejtomcik — CHANGES_REQUESTED** [review]

File: coap-gateway/service/log.go (hunk 5 above)
Comment (jkralik) [reply coap-gateway/service/log.go]: Fixed

File: coap-gateway/service/log.go (hunk 6 above)
Comment (jkralik) [reply coap-gateway/service/log.go]: Fixed

File: charts/plgd-hub/README.md (hunk 4 above)
Comment (jkralik) [reply charts/plgd-hub/README.md:37]: I don't know . I just call make in `charts/plgd` directory This need to be checked by @rafajpet

Comment (ondrejtomcik) [reply charts/plgd-hub/README.md:37]: Actually, I removed it in the helm chart and didn't regenerate markdown. Why I removed it - scopes are defined in the authorization code request, not when exchanging the code for the token.

**Review by ondrejtomcik — APPROVED** [review]
